resource-limits

daiagi opened the pull request:
see the README file
it decribes it pretty well

but in short:

Relay server with configurable usage limits via CLI options
* limits are completly and easyily tunable via CLI
* `--default-limits` CLI flag sets limits that make sense for our use case (singaling server only)
* Very well tested (if i may say so)


run `npm i` and  `npm run test:integration` to see it in action

daiagi committed:
impl resource limits

daiagi committed:
unit tests

daiagi committed:
integration test: connections limit

daiagi committed:
better cli

daiagi committed:
integration tests

daiagi committed:
improved test output

daiagi committed:
configurable message types

daiagi committed:
move integration tests

daiagi committed:
lint

daiagi committed:
remove comments

daiagi committed:
fix: defaults of allowed message types

daiagi committed:
change defaults

daiagi committed:
update README

daiagi committed:
remove a comment

daiagi committed:
more comment fixes

lukastanisic99 commented on lib/utils.js:
```diff
@@ -0,0 +1,27 @@
+  const hasValue = i < process.argv.length - 1
+  if (!hasValue) return defaultValue
+
+  let value = process.argv[i + 1]
```
would rename this function to `getArgValue` or something similar so it's more intuitive what is should do - especially since it's pure js code :D 

Don't understand the `Boolean` check - you assume if it's a boolean flag that if it's set it's always true? 
`--isSomething false` would return true for `argv("isSomething", Boolean)

If the flag is set I'd first always fetch the value and interpret it later for the specific type

daiagi replied:
Hey, just to clarify:
I moved this function to utils in order to re-use.No changes were made to the logic itself.
Regarding the CLI behavior: yes, it's normal for boolean flags to be `true` when present. This code fololws the common pattern:


```bash
myApp --something # true
myApp             #false
```

If we ever need to explicitly turn things off, the typical alternatives are:

```bash
myApp --enable-something
myApp --disable-something
```

Concretely, there are two usages in the code:

```js
const behindProxy = argv('behind-proxy', Boolean)
const useDefaults = argv('default-limits', Boolean)
```

in both cases, supplying the flag means true

If we ever find a need to do something like --isSomething false(which is a bit awkward in CLI design), we can revisit. But for now, this design works well and doesn’t need to be changed.

lukastanisic99 commented on lib/resource-manager.js:
```diff
@@ -0,0 +1,290 @@
+    //   totalDataUsage: number,        // Total bytes across all connections
+    //   activeConnections: Set<string> // Set of connection aliases
+    // }>
+    this.usage = new Map()
```
Maybe not the best to map `users` to `public keys` since they can have an infinite number of them. Probably better to map it against the user's `ip` and/or `ip:port` (if users from the same ISP are playing and the ISP allocates them the same `public ip` in the NAT table) - In general this is also easy to circumvent, since someone can connect to a VPN to get a new ip. 

Some ideas that I'll throw out:
Since we're not exclusively a p2p protocol, we are dependent on web3 and our systems are designed to be used with `assets` that have value, otherwise no point on using blockchains - we can map the connection to on-chain addresses and allow the connection only if the address has 'assets' which is harder to spam since it requires allocating capital. Downside of this is that it would require the relayer also to query the RPC node, which adds more dependencies. Also a handshake `onConnection` would be required between the relayer and the peer, where the peer proves ownership of the private key to the address they're claiming to be - something similar what we already have in the SDK 

In general this is maybe the hardest problem to solve here - we should make this polymorphic so we can support different mappings for `userId`

lukastanisic99 replied:
Also is it worth exploring [cloudflare ](https://www.cloudflare.com/) against standard ddos attacks?

lukastanisic99 commented on lib/resource-manager.js:
```diff
@@ -0,0 +1,290 @@
+    this.usage = new Map()
+
+    // Track per-connection data usage
+    // Map<connectionAlias: string, {
```
What's the `connectionAlias`? The `nextId()` that's incremented for each connection?

lukastanisic99 reviewed (approved):
Overall really good. Wish the repository was in typescript, which is 10x easier to review given that you don't have to dig through the code to find what a variable is.

lukastanisic99 reviewed (changes requested):
Didn't notice that some tests are not passing - result is `not ok` and need to run `npx standard --fix` initially to format the files for tests + some error out totally 
<img width="899" alt="image" src="https://github.com/user-attachments/assets/c312da3f-4477-4b92-a7ad-a73176108f5c" />

daiagi commented:
Hey @lukastanisic99
The failing test is not from my code. You can see it failes on main branch too

Maybe it's part of the reason why the latest version didn't work for you.

I'd suggest just moving the new code to the version that works for us

Or, I could make an issue or try a pr into holepunch's upstream repo 

As for the other comments

I like the name argv. Plus it was there before

The identifying key can be anything. Public key was just the name in the repo, it could be IP, blockchain address or whatever.

lukastanisic99 commented:
> Hey @lukastanisic99 The failing test is not from my code. You can see it failes on main branch too
> 
> Maybe it's part of the reason why the latest version didn't work for you.
> 
> I'd suggest just moving the new code to the version that works for us
> 
> Or, I could make an issue or try a pr into holepunch's upstream repo
> 
> As for the other comments
> 
> I like the name argv. Plus it was there before
> 
> The identifying key can be anything. Public key was just the name in the repo, it could be IP, blockchain address or whatever.

Yeah the error code is failing on main too, but new tests for me are returning `not ok` 

<img width="833" alt="image" src="https://github.com/user-attachments/assets/c07727a2-e1e4-45e8-86e8-4c5f194f5f1d" />


As for `argv`, the name isn't that important you can keep it. I just don't understand their boolean logic as you can only have `true` flags if used.

As for the `key`, I know the key can be anything, but it should be easy for us to swap strategies for what the `key` is, but that can be another PR

daiagi committed:
lint + identity strategy

daiagi committed:
remove console logs

daiagi commented:
* Linted the code and added the npm run lint command.
* Added the identity strategy.
* Discovered a bug where if the client started multiple Node instances, resource protection would apply per instance—creating a potential attack vector.
* Fixed the bug.

#### Background

* Previously, a new `ResourceManager` was created inside the `NodeProxy` constructor:
```js
Client: new Node(stream) → sends handshake
Server: protocol.once('handshake', onHandshake) → new NodeProxy → new ResourceManager
```

This meant a client could spin up multiple `Node` instances (index.js), Each client connection would trigger the server to create a separate NodeProxy with its own ResourceManager. As a result, resource limits could be bypassed by creating multiple client connections.

#### Fix
Now, a single ResourceManager is created at server startup (in bin.js) and passed to the relay() function. All NodeProxy instances share this instance. It's essentially a singleton, but instead of being globally scoped, it's passed around as a function argument—making it easier for tests and dependency-inject


```js
// bin.js
const sharedResourceManager = new ResourceManager(resourceLimits)
relay(node, new Stream(false, socket), sharedResourceManager, options)

// index.js relay function  
protocol.once('handshake', (message) => {
  new NodeProxy(dht, protocol, keyPair, sharedResourceManager, options) // ← shared!
})
```

daiagi committed:
lint

daiagi committed:
fix tests

daiagi committed:
fix test

daiagi commented:
ok. now this branch has exactly the same test fail as the main branch

i tried briefly to fix the main branch test, but it is quite complicated

and it has something to do with custody over the crypto keys
- https://github.com/holepunchto/hyperswarm-dht-relay/issues/26

momoshell commented on lib/node-proxy.js:
```diff
@@ -88,18 +112,32 @@ function onStreamClose () {
+    return
+  }
+
+  this._resourceManager.recordConnection(clientId, remoteAlias)
```
`recordConnection` is keyed by `remoteAlias` (= `message.alias`, generated by each client's own `nextId()` starting at 1 — see `lib/node.js:87`), not by the relay-unique `alias` computed at line 116. 

`ResourceManager` is a single instance shared across all sockets (`bin.js:35`), so `connectionData` is one process-wide map keyed by a value every client independently produces starting at 1, 2, 3…

Two unrelated clients both opening a first connection collide on alias `1`. Client A's `connectionData[1]` gets overwritten by client B's.
Data gets billed to the wrong client, and when A disconnects, `recordDisconnection` deletes `connectionData[1]`, which is now B's live connection, so B's next `onData` hits `CONNECTION_NOT_FOUND` and gets destroyed. Any client can knock out another client's connection just by reusing a low alias number, which happens by default.

Suggest keying by the relay-generated `alias` instead, or `` `${clientId}:${remoteAlias}` ``.

This also answers the `connectionAlias` question from the other thread — it's the client-chosen `remoteAlias`, not the relay's `nextId()`.

momoshell commented on lib/node-proxy.js:
```diff
@@ -119,6 +159,8 @@ function onConnect (message) {
   }
 
   const onClose = () => {
+    this._resourceManager.recordDisconnection(clientId, remoteAlias)
```
`onError` (line 151) and `onClose` both call `recordDisconnection` for the same connection, and `stream.off('error', onError)` only happens inside `onClose`, so an errored stream fires both handlers, decrementing the counter twice for one connection.

Cycling connections to an unreachable peer drives a client's counted `connections` toward 0 while real connections stay open, defeating `maxConnections`. Suggest only recording the disconnection once per connection (e.g. only from `onClose`, which always fires, or guard with a `settled` flag).

momoshell commented on lib/resource-manager.js:
```diff
@@ -0,0 +1,278 @@
+
+    if (connectionAlias) {
+      client.activeConnections.delete(connectionAlias)
+      this.connectionData.delete(connectionAlias)
```
`recordDisconnection` decrements `client.connections` but never removes the `usage` entry when it reaches 0, and there's no cap on the number of distinct `clientId`s tracked. With the default `publicKey` strategy, `clientId` is client-supplied per connect message at zero cost. 

Cycling a fresh keypair per connection leaves a permanent `usage` entry forever. That's an unbounded-memory-growth vector independent of whether the connection/data limits themselves are working correctly.

momoshell commented on lib/resource-manager.js:
```diff
@@ -0,0 +1,278 @@
+    return true
+  }
+
+  checkMessage (_clientId, messageType, _messageData = null) {
```
`checkMessage` is never called anywhere in `lib/`, `bin.js`, or `index.js` (checked via grep), so `--allowed-message-types` currently enforces nothing.
